components/function-controller/internal/controllers/serverless/build_resources.go

```diff
@@ -172,6 +189,118 @@ func (r *FunctionReconciler) buildService(log logr.Logger, instance *serverlessv
 			},
 		},
 	}
+}
+
+func (r *FunctionReconciler) buildInternalImageAddress(instance *serverlessv1alpha1.Function) string {
+	imageTag := r.calculateImageTag(instance)
+	return fmt.Sprintf("%s/%s-%s:%s", r.config.Docker.Address, instance.Namespace, instance.Name, imageTag)
+}
+
+func (r *FunctionReconciler) buildExternalImageAddress(instance *serverlessv1alpha1.Function) string {
+	imageTag := r.calculateImageTag(instance)
+	return fmt.Sprintf("%s/%s-%s:%s", r.config.Docker.ExternalAddress, instance.Namespace, instance.Name, imageTag)
+}
+
+func (r *FunctionReconciler) sanitizeDependencies(dependencies string) string {
+	result := "{}"
+	if strings.Trim(dependencies, " ") != "" {
+		result = dependencies
+	}
+
+	return result
+}
+
+func (r *FunctionReconciler) functionLabels(instance *serverlessv1alpha1.Function) map[string]string {
+	labels := make(map[string]string, len(instance.GetLabels())+3)
+	for key, value := range instance.GetLabels() {
+		labels[key] = value
+	}
+
+	labels[serverlessv1alpha1.FunctionNameLabel] = instance.Name
+	labels[serverlessv1alpha1.FunctionManagedByLabel] = "function-controller"
+	labels[serverlessv1alpha1.FunctionUUIDLabel] = string(instance.GetUID())
+
+	return labels
+}
+
+func (r *FunctionReconciler) serviceLabels(instance *serverlessv1alpha1.Function) map[string]string {
+	serviceLabels := r.functionLabels(instance)
+	serviceLabels[servingKnativeVisibilityLabel] = servingKnativeVisibilityLabelValue
+	return serviceLabels
+}
+
+func (r *FunctionReconciler) servicePodAnnotations(instance *serverlessv1alpha1.Function) map[string]string {
+	annotations := map[string]string{
+		autoscalingKnativeMinScaleAnn: "1",
+		autoscalingKnativeMaxScaleAnn: "1",
+	}
+	if instance.Spec.MinReplicas != nil {
+		annotations[autoscalingKnativeMinScaleAnn] = fmt.Sprintf("%d", *instance.Spec.MinReplicas)
+	}
+	if instance.Spec.MaxReplicas != nil {
+		annotations[autoscalingKnativeMaxScaleAnn] = fmt.Sprintf("%d", *instance.Spec.MaxReplicas)
+	}
+	return annotations
+}
+
+func (r *FunctionReconciler) servicePodLabels(log logr.Logger, instance *serverlessv1alpha1.Function, oldService *servingv1.Service) map[string]string {
+	functionLabels := r.functionLabels(instance)
+	bindingLabels := r.retrieveBindingLabels(log, instance, oldService)
+	podLabels := instance.Spec.PodLabels
+
+	if podLabels == nil || len(podLabels) == 0 {
+		for key, value := range bindingLabels {
+			functionLabels[key] = value
+		}
+		return functionLabels
+	}
+
+	for key, value := range functionLabels {
+		podLabels[key] = value
+	}
+	for key, value := range bindingLabels {
+		podLabels[key] = value
+	}
+
+	return podLabels
+}
```

Comment:
Instead of having a special if for empty podLabels map, better is to create a new map that you will return - code is much simpler
```suggestion
func (r *FunctionReconciler) servicePodLabels(log logr.Logger, instance *serverlessv1alpha1.Function, oldService *servingv1.Service) map[string]string {
	functionLabels := r.functionLabels(instance)
	bindingLabels := r.retrieveBindingLabels(log, instance, oldService)
	podLabels := instance.Spec.PodLabels

	// len(functionLabels) as we cannot predict the exact size of map
	result := make(map[string]string, len(functionLabels))
	for key, value := range podLabels {
		result[key] = value
	}
	for key, value := range functionLabels {
		result[key] = value
	}
	//TODO: remove after migration of Bindings to Function CR
	for key, value := range bindingLabels {
		result[key] = value
	}

	return result
}
```

https://play.golang.org/p/wbjfV82CXvH

Reply:
Also, it is not good to update `podLabels` because you updating `instance.Spec.PodLabels` field in such a case and some logic that uses that field may not work as expected. It is a good practice to not modify arguments of function inside the function.

Reply:
Hm... you can even create a method responsible for merging labels, as we do that in many places, something like that:

```go
func mergeLabels(labels ...map[string]string) map[string]string {
	result := make(map[string]string, 0)
	for _, labelsMap := range labels {
		for key, value := range labelsMap {
			result[key] = value
		}
	}

	return result
}
```
https://play.golang.org/p/Ud9XM3SPXMF